mm2src/mm2_bitcoin/keys/src/address/address_builder.rs

```diff
@@ -0,0 +1,110 @@
+use crypto::ChecksumType;
+use {Address, AddressFormat, AddressHashEnum, AddressPrefixes, AddressScriptType, NetworkAddressPrefixes};
+
+/// Params for AddressBuilder to select output script type
+#[derive(PartialEq)]
+pub enum AddressBuilderOption {
+    /// build for pay to pubkey hash output (witness or legacy)
+    BuildPubkeyHash,
+    /// build for pay to script hash output (witness or legacy)
+    BuildScriptHash,
+}
+
+/// Builder for Address
+/// Returns Address struct depending on addr_format value checking other properties
+pub struct AddressBuilder {
+    /// Coin base58 address prefixes from coin config
+    pub prefixes: NetworkAddressPrefixes,
+    /// Segwit addr human readable part
+    pub hrp: Option<String>,
+    /// Public key hash.
+    pub hash: AddressHashEnum,
+    /// Checksum type
+    pub checksum_type: ChecksumType,
+    /// Address Format
+    pub addr_format: AddressFormat,
+}
+
+impl AddressBuilder {
+    /// Builds Address with the script type corresponding to p2pkh or p2wpkh
+    pub fn build_as_pkh(&self) -> Result<Address, String> { self.build(AddressBuilderOption::BuildPubkeyHash) }
+
+    /// Builds Address with the script type corresponding to p2sh or p2wsh
+    pub fn build_as_sh(&self) -> Result<Address, String> { self.build(AddressBuilderOption::BuildScriptHash) }
+
+    pub fn build(&self, build_option: AddressBuilderOption) -> Result<Address, String> {
```

Comment:
To follow the builder pattern, `build_option: AddressBuilderOption` should be part of `AddressBuilder` struct and you should add either a function `pub fn with_options(mut self, build_option: AddressBuilderOption)` or 2 functions instead, `fn as_pkh(mut self)` and `as_sh(mut self)` that take a mut self too. We should also have `new` function and other setter functions for other fields.

ref. https://rust-unofficial.github.io/patterns/patterns/creational/builder.html

Reply:
fixed in [64dbf87](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1960/commits/64dbf873bbcc5a92d86e3bc107a9ad7d6db3b715)